core/shared/src/main/scala/zio/Config.scala

```diff
@@ -0,0 +1,341 @@
+    }
+  }
+  case object Secret extends Atom[zio.Secret] {
+    final def parse(text: String): Either[Config.Error, zio.Secret] = Right(
+      zio.Secret(Chunk.fromIterable(text.getBytes()))
+    )
+  }
+  final case class Sequence[A](config: Config[A])   extends Composite[Chunk[A]]
+  final case class Table[V](valueConfig: Config[V]) extends Composite[Map[String, V]]
+  case object Text extends Atom[String] {
+    final def parse(text: String): Either[Config.Error, String] = Right(text)
+  }
+  final case class Zipped[A, B, C](left: Config[A], right: Config[B], zippable: Zippable.Out[A, B, C])
+      extends Composite[C]
+
+  /**
+   * The possible ways that loading configuration data may fail.
+   */
+  sealed trait Error extends Exception with NoStackTrace { self =>
+    def &&(that: Error): Error = Error.And(self, that)
+    def ||(that: Error): Error = Error.And(self, that)
+
+    def prefixed(prefix: Chunk[String]): Error
+
+    override def getMessage(): String = toString()
+  }
+  object Error {
+    final case class And(left: Error, right: Error) extends Error {
+      def prefixed(prefix: Chunk[String]): And =
+        copy(left = left.prefixed(prefix), right = right.prefixed(prefix))
+
+      override def toString(): String = s"(${left.toString()} and ${right.toString()})"
+    }
+    final case class InvalidData(path: Chunk[String], message: String) extends Error {
+      def prefixed(prefix: Chunk[String]): InvalidData = copy(path = prefix ++ path)
+
+      override def toString(): String = s"(Invalid data at ${path.mkString(".")}: ${message})"
+    }
+    final case class MissingData(path: Chunk[String], message: String) extends Error {
+      def prefixed(prefix: Chunk[String]): MissingData = copy(path = prefix ++ path)
+
+      override def toString(): String = s"(Missing data at ${path.mkString(".")}: ${message})"
+    }
+    final case class Or(left: Error, right: Error) extends Error {
+      def prefixed(prefix: Chunk[String]): Or =
+        copy(left = left.prefixed(prefix), right = right.prefixed(prefix))
+
+      override def toString(): String = s"(${left.toString()} or ${right.toString()})"
+    }
+    final case class SourceUnavailable(path: Chunk[String], message: String, cause: Cause[Throwable]) extends Error {
+      def prefixed(prefix: Chunk[String]): SourceUnavailable = copy(path = prefix ++ path)
+
+      override def toString(): String = s"(Source unavailable at ${path.mkString(".")}: ${message})"
+    }
+    final case class Unsupported(path: Chunk[String], message: String) extends Error {
+      def prefixed(prefix: Chunk[String]): Unsupported = copy(path = prefix ++ path)
+
+      override def toString(): String = s"(Unsupported operation at ${path.mkString(".")}: ${message})"
+    }
+  }
+
+  def bigDecimal: Config[BigDecimal] = Decimal
+
+  def bigDecimal(name: String): Config[BigDecimal] = bigDecimal.nested(name)
+
+  def bigInt: Config[BigInt] = Integer
```

Comment:
Don't they need to be wrapped in `LazyConfig`?

Reply:
No, not necessary because there are no side-effects or anything. 

However, I see I failed to make binary operators lazy on the RHS. I made that fix now.

Reply:
Don't we need it for even zero side-effect ones as well to handle recursive configs?

Reply:
Or may be `Recursion` is another node and it is unto the interpreter to handle it.

Reply:
Makes sense